### DISCUSSION TITLE
SMTP - Self-signed Certificate in Certificate Chain

### DISCUSSION THREAD

**Discussion by kubowler99** [directus/directus#16857] [Archive]
All - Looking for help. Has anyone solved the self-signed cert in certificate chain error with SMTP emails? 

Error being returned by Directus:

"message": "self-signed certificate in certificate chain",
"stack": "Error: self-signed certificate in certificate chain

I've confirmed using native nodemailer with the following configuration works:

    host: "exampleHost",
    port: 587,
    secure: false, // true for 465, false for other ports
    ignoreTLS: false,
    requireTLS: true,
    auth: {
        user: 'mailUsername',
        pass: 'mailPassword',
    },
    tls: {
        rejectUnauthorized: false
    }

But, when setting the following configuration properties in directus, I still see the cert error:

    EMAIL_SMTP_SECURE: false,
    EMAIL_SMTP_IGNORE_TLS: false,
    EMAIL_SMTP_REQUIRE_TLS: true,
    EMAIL_SMTP_TLS__REJECT_UNAUTHORIZED: false,
    EMAIL_SMTP_USER: mailUsername,
    EMAIL_SMTP_PASSWORD: mailPassword

Is there something I'm potentially missing in the config? Has anyone else used

EMAIL_SMTP_TLS__REJECT_UNAUTHORIZED: false

successfully?

**Comment by paescuj** [comment]
~`EMAIL_SMTP_TLS__REJECT_UNAUTHORIZED` is (currently) not supported. See here for all supported configs: https://docs.directus.io/self-hosted/config-options.html#smtp-smtp~

**Edit:** Just took a look at the code and it turns out it is supported by using `EMAIL_SMTP_TLS_REJECT_UNAUTHORIZED` (only single underscore).

**Reply by kubowler99** [reply]
That was it! Thanks for the assist!

**Comment by ddelange** [comment]
(cross posting here because the other SELF_SIGNED_CERT_IN_CHAIN directus issue is locked)

## Problem

Managed databases like Heroku Managed PostgreSQL or DigitalOcean Managed PostgreSQL generally enforce SSL, and provide a **self-signed** CA certificate so you can configure [PostgreSQL SSL](https://www.postgresql.org/docs/current/libpq-ssl.html#:~:text=Table%C2%A032.1.%C2%A0SSL%20Mode%20Descriptions) using `?sslmode=verify-full`.

When using node.js (knex uses [node-postgres](https://node-postgres.com/) uses [node.js SSL](https://nodejs.org/api/tls.html)), self-signed CA certificates are rejected by default with an error similar to:

```
Error: self-signed certificate in certificate chain
    at TLSSocket.onConnectSecure (node:_tls_wrap:1679:34)
    at TLSSocket.emit (node:events:519:28)
    at TLSSocket._finishInit (node:_tls_wrap:1078:8)
    at ssl.onhandshakedone (node:_tls_wrap:864:12) {
  code: 'SELF_SIGNED_CERT_IN_CHAIN'
}
```

Even when specifying env vars like:

```sh
DB_CONNECTION_STRING="postgresql://...?sslmode=verify-full" # sslmode param not parsed by node-postgres (it's a libpg/psql thing)
DB_SSL__REJECT_UNAUTHORIZED="true"                          # verify-full behaviour in node-postgres
DB_SSL__CA=<multiline-ca-cert-in-pem-format>                # CA certificate as string
DB_SSL__CA_FILE="/absolute/path/to/pg-ca.crt"               # CA certificate as absolute path (preferred)
```

## Solution 

The [solution](https://stackoverflow.com/a/47160447/5511061) is to make node.js trust this specific self-signed CA certificate using `NODE_EXTRA_CA_CERTS`. Provide the absolute path to the database's self-signed CA certificate in PEM format so it gets added to the node.js trust store:

```
NODE_EXTRA_CA_CERTS="/absolute/path/to/pg-ca.crt"
```

## Warning

⚠️ Never set `NODE_TLS_REJECT_UNAUTHORIZED` / `DB_SSL__REJECT_UNAUTHORIZED` / `rejectUnauthorized` to false in production: **it will compromise security of your entire application!**

> *Originally posted at https://stackoverflow.com/a/79807048/5511061*

**Reply by ddelange** [reply]
@rijkvanzanten could you maybe cross post this to the relevant issues for visibility?

https://github.com/directus/directus/issues/24471
https://github.com/directus/directus/issues/20689
https://github.com/directus/directus/issues/3314
https://github.com/directus/directus/issues/3267

I wasted hours on this, despite finding these relevant issues.

**Reply by jacobsandersen** [reply]
Oh my god. Finally this `NODE_EXTRA_CA_CERTS` worked. I've been at this for hours with various different attempts.

This needs to be called out when using a self signed cert with own CA.

cc @rijkvanzanten

**Reply by ComfortablyCoding** [reply]
@jacobsandersen Happy to take a PR (or issue with findings/notes etc) to the [docs](https://github.com/directus/docs) with any clarifications or workflows needed for utilizing self signed certs.

**Reply by jacobsandersen** [reply]
> @jacobsandersen Happy to take a PR (or issue with findings/notes etc) to the [docs](https://github.com/directus/docs) with any clarifications or workflows needed for utilizing self signed certs.

I'll see if I can make a PR later.

**Reply by rijkvanzanten** [reply]
@jacobsandersen I happen to be doing some work on the docs as we speak. Where would you expect to find this in the docs?

**Reply by ddelange** [reply]
I would expect my above explanation about managed databases providing self-signed certs somewhere in the [database docs](https://directus.io/docs/configuration/database)

**Reply by jacobsandersen** [reply]
Agree with @ddelange. Somewhere in the database docs, maybe with the env vars.
